runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java

```diff
@@ -496,13 +498,24 @@ public void add(CompletionStage<WindowedValue<OutT>> element) {
       checkState(
           !collectorSealed.get(),
           "Cannot add elements to an unprepared collector. Make sure prepare() is invoked before adding elements.");
-      outputFutures.add(element);
+      outputFuture =
+          outputFuture.thenCombine(
+              element,
+              (collection, event) -> {
+                collection.add(event);
+                return collection;
+              });
+    }
+
+    @Override
+    public void addAll(CompletionStage<Collection<WindowedValue<OutT>>> elements) {
+      outputFuture = FutureUtils.combineFutures(outputFuture, elements);
```

Comment:
Can we add the state check to ensure `addAll(...)` isn't invoked when the collector is sealed?

Reply:
Yup, added.